File: MIDTERM_LAB_3/programming-2/assignment-02/MP06.java
/**
 * =====================================================
 * Student Name    : UNDUG, MOHAMMAD FAISSAL A.
 * Course          : Programming 2
 * Assignment      : MP06 - Display Unique Values
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 18, 2026
 * GitHub Repo     : https://github.com/[your-username]/uphsd-cs-undug-mohammadfaissal
 *
 * Description:
 *   This program reads a CSV dataset file whose path is entered by the user.
 *   It parses each row and collects all unique values found in the "Exam"
 *   column, then prints them in a numbered list with a total count at the end.
 * =====================================================
 */

import java.io.*;
import java.util.*;

public class MP06 {

    // -- SECTION 1: Entry Point --------------------------------------------------
    // Main method prompts the user for the CSV file path, then calls the
    // processing method to find and display all unique exam types.
    public static void main(String[] args) throws IOException {

        // Ask the user to type in the path to the CSV file
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the path to the CSV file: ");
        String filePath = scanner.nextLine().trim();

        // Pass the file path to the method that handles the unique value logic
        displayUniqueExams(filePath);

        scanner.close();
    }

    // -- SECTION 2: CSV Reader and Unique Value Collector ------------------------
    // Opens the CSV file, skips the header rows, then reads each data row.
    // For every row, it extracts the "Exam" column (index 3) and adds it to
    // a TreeSet which automatically removes duplicates and sorts alphabetically.
    static void displayUniqueExams(String filePath) throws IOException {

        // TreeSet keeps values sorted and automatically ignores duplicates
        Set<String> uniqueExams = new TreeSet<>();

        // Open the file using a BufferedReader for efficient line-by-line reading
        BufferedReader reader = new BufferedReader(new FileReader(filePath));

        String line;
        int lineNumber = 0;

        // Read through every line in the file
        while ((line = reader.readLine()) != null) {
            lineNumber++;

            // Skip the first 6 lines which are metadata/header rows, not data
            if (lineNumber <= 6) continue;

            // Skip the column header row (line 7)
            if (lineNumber == 7) continue;

            // Skip blank lines to avoid empty entries
            if (line.trim().isEmpty()) continue;

            // Parse the CSV row into individual fields, handling quoted commas
            String[] fields = parseCSVLine(line);

            // The Exam column is at index 3 (0-based)
            // Make sure the row has enough columns before accessing index 3
            if (fields.length > 3) {
                String exam = fields[3].trim();
                if (!exam.isEmpty()) {
                    uniqueExams.add(exam);
                }
            }
        }

        reader.close();

        // -- SECTION 3: Output ---------------------------------------------------
        // Print all unique exam types in a numbered, readable format
        System.out.println();
        System.out.println("=".repeat(55));
        System.out.println("  MP06 - UNIQUE EXAM TYPES IN DATASET");
        System.out.println("  Student: UNDUG, MOHAMMAD FAISSAL A.");
        System.out.println("=".repeat(55));
        System.out.println();

        int count = 1;
        // Loop through the sorted set and print each unique exam name
        for (String exam : uniqueExams) {
            System.out.printf("  %2d. %s%n", count++, exam);
        }

        System.out.println();
        System.out.println("=".repeat(55));
        System.out.printf("  Total unique exam types found: %d%n", uniqueExams.size());
        System.out.println("=".repeat(55));
    }

    // -- SECTION 4: CSV Line Parser ----------------------------------------------
    // Splits a single CSV line into fields while correctly handling values
    // that are wrapped in double quotes and may contain commas inside them.
    // For example: "Undug,Mohammad" should be treated as ONE field, not two.
    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean insideQuotes = false;

        // Go through each character one by one
        for (char c : line.toCharArray()) {
            if (c == '"') {
                // Toggle the insideQuotes flag when we encounter a quote mark
                insideQuotes = !insideQuotes;
            } else if (c == ',' && !insideQuotes) {
                // A comma outside quotes signals the end of a field
                fields.add(current.toString());
                current.setLength(0); // Reset the buffer for the next field
            } else {
                // Any other character just gets added to the current field
                current.append(c);
            }
        }

        // Add the last field after the loop ends
        fields.add(current.toString());
        return fields.toArray(new String[0]);
    }
}

File: MIDTERM_LAB_3/programming-2/assignment-02/mp06.js
/**
 * =====================================================
 * Student Name    : UNDUG, MOHAMMAD FAISSAL A.
 * Course          : Programming 2
 * Assignment      : MP06 - Display Unique Values
 * School          : University of Perpetual Help System DALTA, Molino Campus
 * Date            : March 18, 2026
 * GitHub Repo     : https://github.com/[your-username]/uphsd-cs-undug-mohammadfaissal
 * Runtime         : Node.js (run with: node mp06.js)
 *
 * Description:
 *   This program reads a CSV dataset file whose path is entered by the user.
 *   It parses each row and collects all unique values found in the "Exam"
 *   column, then prints them in a numbered list with a total count at the end.
 * =====================================================
 */

"use strict";

// Built-in Node.js modules for file reading and user input
const fs            = require("fs");
const readline      = require("readline");

// -- SECTION 1: CSV Line Parser -----------------------------------------------
// Splits one CSV line into an array of field strings.
// Handles fields wrapped in double quotes that may contain commas inside them.
// Iterates character by character, toggling insideQuotes when a " is found,
// and only treating a comma as a field separator when outside quotes.
function parseCSVLine(line) {
    const fields = [];
    let current = "";
    let insideQuotes = false;

    for (const c of line) {
        if (c === '"') {
            // Flip the quote flag — we entered or exited a quoted section
            insideQuotes = !insideQuotes;
        } else if (c === "," && !insideQuotes) {
            // Comma outside quotes = end of one field, start of next
            fields.push(current);
            current = "";
        } else {
            current += c;
        }
    }

    // Push the last field that has no trailing comma
    fields.push(current);
    return fields;
}

// -- SECTION 2: Unique Exam Extractor -----------------------------------------
// Reads all lines from the file, skips the first 7 (metadata + header),
// parses each data row, and adds the Exam field (index 3) to a Set.
// A Set automatically discards duplicate values, leaving only unique ones.
// The resulting set is converted to a sorted array for display.
function displayUniqueExams(filePath) {

    // Read the entire file as a UTF-8 string, then split into individual lines
    const rawText = fs.readFileSync(filePath, "utf8");
    const lines   = rawText.split(/\r?\n/); // Handle both Windows and Unix line endings

    // Use a Set to collect exam names — duplicates are ignored automatically
    const uniqueExams = new Set();

    // Lines 0-6 (first 7) are metadata/headers — start reading data from index 7
    for (let i = 7; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue; // Skip blank lines

        const fields = parseCSVLine(line);

        // Exam is at column index 3; only add if the field has content
        if (fields.length > 3 && fields[3].trim()) {
            uniqueExams.add(fields[3].trim());
        }
    }

    // Convert Set to array and sort alphabetically for clean output
    const sortedExams = [...uniqueExams].sort();

    // -- SECTION 3: Output ------------------------------------------------------
    // Print all unique exam names in a numbered list
    const line55 = "=".repeat(55);
    console.log();
    console.log(line55);
    console.log("  MP06 - UNIQUE EXAM TYPES IN DATASET");
    console.log("  Student: UNDUG, MOHAMMAD FAISSAL A.");
    console.log(line55);
    console.log();

    sortedExams.forEach((exam, index) => {
        // Pad the number so single and double digits align neatly
        console.log(`  ${String(index + 1).padStart(2)}. ${exam}`);
    });

    console.log();
    console.log(line55);
    console.log(`  Total unique exam types found: ${sortedExams.length}`);
    console.log(line55);
}

// -- SECTION 4: Program Entry Point -------------------------------------------
// Uses readline to prompt the user for the CSV file path in the terminal.
// Once the user presses Enter, the path is passed to the main processing function.
const rl = readline.createInterface({
    input:  process.stdin,
    output: process.stdout
});

rl.question("Enter the path to the CSV file: ", (filePath) => {
    rl.close();
    displayUniqueExams(filePath.trim());
});
